File: cpp/Classes/TurnBasedMultiplayerScene.cpp
#include "TurnBasedMultiplayerScene.hpp"
#include "Utils.h"
#include "RealTimeMultiplayerScene.hpp"

#if CC_TARGET_PLATFORM == CC_PLATFORM_ANDROID
#include "NearbyConnectionsScene.hpp"
#endif

using namespace cocos2d;

cocos2d::Scene *TurnBasedMultiplayerScene::createScene()
{
    auto scene = Scene::create();

    auto layer = TurnBasedMultiplayerScene::create();

    scene->addChild(layer);

    return scene;
}

bool TurnBasedMultiplayerScene::init()
{
    if(!BaseScene::init())
    {
        return false;
    }

    Size size = Director::getInstance()->getWinSize();
    std::string defaultFont("arial.ttf");

    setTitle("Turn Based Multiplayer");

    Menu* menu = Menu::create(
                              MenuItemFont::create("Create TurnBased Match", CC_CALLBACK_1(TurnBasedMultiplayerScene::CreateTurnBasedMatch, this)),
                              MenuItemFont::create("Show Match Inbox UI", CC_CALLBACK_1(TurnBasedMultiplayerScene::ShowMatchInboxUI, this)),
//                              MenuItemFont::create("Accept Invitation", CC_CALLBACK_1(TurnBasedMultiplayerScene::AcceptInvitation, this)),
//                              MenuItemFont::create("Decline Invitation", CC_CALLBACK_1(TurnBasedMultiplayerScene::DeclineInvitation, this)),
//                              MenuItemFont::create("Dismiss Invitation", CC_CALLBACK_1(TurnBasedMultiplayerScene::DismissInvitation, this)),
                              MenuItemFont::create("Dismiss Match", CC_CALLBACK_1(TurnBasedMultiplayerScene::DismissMatch, this)),
                              MenuItemFont::create("Fetch Match", CC_CALLBACK_1(TurnBasedMultiplayerScene::FetchMatch, this)),
                              MenuItemFont::create("Take My Turn", CC_CALLBACK_1(TurnBasedMultiplayerScene::TakeMyTurn, this)),
                              MenuItemFont::create("Finish Match During MyTurn", CC_CALLBACK_1(TurnBasedMultiplayerScene::FinishMatchDuringMyTurn, this)),
                              MenuItemFont::create("Rematch", CC_CALLBACK_1(TurnBasedMultiplayerScene::Rematch, this)),
                              MenuItemFont::create("Leave Match", CC_CALLBACK_1(TurnBasedMultiplayerScene::LeaveMatch, this)),
                              MenuItemFont::create("CancelMatch", CC_CALLBACK_1(TurnBasedMultiplayerScene::CancelMatch, this)),
                              MenuItemFont::create("FetchMatches", CC_CALLBACK_1(TurnBasedMultiplayerScene::FetchMatches, this)),
                              MenuItemFont::create("Show Player Select UI", CC_CALLBACK_1(TurnBasedMultiplayerScene::ShowPlayerSelectUI, this)),
                              nullptr
                              );

    menu->alignItemsVerticallyWithPadding(5);
    menu->setPosition(size.width/2, size.height/2 + 40);
    addChild(menu);

    return true;
}

void TurnBasedMultiplayerScene::onPrevScene(cocos2d::Ref *sender)
{
    Director::getInstance()->replaceScene(RealTimeMultiplayerScene::createScene());
}

void TurnBasedMultiplayerScene::onNextScene(cocos2d::Ref *sender)
{
#if CC_TARGET_PLATFORM == CC_PLATFORM_ANDROID
    Director::getInstance()->replaceScene(NearbyConnectionsScene::createScene());
#else
    BaseScene::onNextScene(sender);
#endif
}

void TurnBasedMultiplayerScene::CreateTurnBasedMatch(cocos2d::Ref *sender) {
    auto config = gpg::TurnBasedMatchConfig::Builder()
    .SetMaximumAutomatchingPlayers(2)
    .SetMinimumAutomatchingPlayers(1)
    .Create();
    _game_services->TurnBasedMultiplayer()
    .CreateTurnBasedMatch(config,
                          [this](gpg::TurnBasedMultiplayerManager::TurnBasedMatchResponse const & response) {
                              if (gpg::IsSuccess(response.status)) {
                                  CCLOG("create turn based match success");
                                  this->_match = response.match;
                                  this->_myId = _match.PendingParticipant().Id();
                                  this->handleMatch(response.match);
                              } else {
                                  CCLOG("create turn based match failed:%d", response.status);
                              }
                          });
}

void TurnBasedMultiplayerScene::ShowMatchInboxUI(cocos2d::Ref *sender) {
    _game_services->TurnBasedMultiplayer()
    .ShowMatchInboxUI([this](gpg::TurnBasedMultiplayerManager::MatchInboxUIResponse const & response) {
        if (gpg::IsSuccess(response.status)) {
            CCLOG("show match inbox success");
            this->_match = response.match;
            this->handleMatch(response.match);
        } else {
            CCLOG("show match inbox failed:%d", response.status);
        }
    });
}

void TurnBasedMultiplayerScene::ShowPlayerSelectUI(cocos2d::Ref *sender) {
    _game_services->TurnBasedMultiplayer()
    .ShowPlayerSelectUI(1,
                        2,
                        true,
                        [](gpg::TurnBasedMultiplayerManager::PlayerSelectUIResponse const & response) {
                            if (gpg::IsSuccess(response.status)) {
                                CCLOG("show player select success");
                            } else {
                                CCLOG("show player select failed:%d", response.status);
                            }
                        });
}

void TurnBasedMultiplayerScene::AcceptInvitation(cocos2d::Ref *sender) {
    gpg::MultiplayerInvitation invitation;
    _game_services->TurnBasedMultiplayer()
    .AcceptInvitation(invitation,
                      [](gpg::TurnBasedMultiplayerManager::TurnBasedMatchResponse const & response) {
                          if (gpg::IsSuccess(response.status)) {
                              CCLOG("accept invitation success");
                          } else {
                              CCLOG("accept invitation failed:%d", response.status);
                          }
                      });
}

void TurnBasedMultiplayerScene::DeclineInvitation(cocos2d::Ref *sender) {
    gpg::MultiplayerInvitation invitation;
    _game_services->TurnBasedMultiplayer().DeclineInvitation(invitation);
}

void TurnBasedMultiplayerScene::DismissInvitation(cocos2d::Ref *sender) {
    gpg::MultiplayerInvitation invitation;
    _game_services->TurnBasedMultiplayer().DismissInvitation(invitation);
}

void TurnBasedMultiplayerScene::DismissMatch(cocos2d::Ref *sender) {
    _game_services->TurnBasedMultiplayer().DismissMatch(_match);
}

void TurnBasedMultiplayerScene::FetchMatch(cocos2d::Ref *sender) {
    if (!_match.Valid()) {
        CCLOG("Match is invalid");
        return;
    }
    _game_services->TurnBasedMultiplayer()
    .FetchMatch(_match.Id(),
                [](gpg::TurnBasedMultiplayerManager::TurnBasedMatchResponse const & resp) {
                    if (gpg::IsSuccess(resp.status)) {
                        CCLOG("fetch match success");
                    } else {
                        CCLOG("fetch match failed:%d", resp.status);
                    }
                });
}

void TurnBasedMultiplayerScene::TakeMyTurn(cocos2d::Ref *sender) {
    if (!_match.Valid()) {
        CCLOG("Match is invalid");
        return;
    }
    
    std::vector<uint8_t> match_data;
    auto result = _match.ParticipantResults().WithResult(_myId, 0, gpg::MatchResult::WIN);
    gpg::MultiplayerParticipant next_participant;
    _game_services->TurnBasedMultiplayer()
    .TakeMyTurn(_match,
                match_data,
                result,
                _match.SuggestedNextParticipant(),
                [](gpg::TurnBasedMultiplayerManager::TurnBasedMatchResponse const &resp) {
                    if (gpg::IsSuccess(resp.status)) {
                        CCLOG("take my turn success");
                    } else {
                        CCLOG("take my turn failed:%d", resp.status);
                    }
                });
}

void TurnBasedMultiplayerScene::FinishMatchDuringMyTurn(cocos2d::Ref *sender) {
    if (!_match.Valid()) {
        CCLOG("Match is invalid");
        return;
    }
    
    std::vector<uint8_t> match_data;
    auto result = _match.ParticipantResults().WithResult(_myId, 0, gpg::MatchResult::WIN);
    _game_services->TurnBasedMultiplayer()
    .FinishMatchDuringMyTurn(_match,
                             match_data,
                             result,
                             [](gpg::TurnBasedMultiplayerManager::TurnBasedMatchResponse const & resp) {
                                 if (gpg::IsSuccess(resp.status)) {
                                     CCLOG("finish match during my turn success");
                                 } else {
                                     CCLOG("finish match during my turn failed:%d", resp.status);
                                 }
                             });
}

void TurnBasedMultiplayerScene::Rematch(cocos2d::Ref *sender) {
    if (!_match.Valid()) {
        CCLOG("Match is invalid");
        return;
    }
    
    _game_services->TurnBasedMultiplayer()
    .Rematch(_match,
             [](gpg::TurnBasedMultiplayerManager::TurnBasedMatchResponse const & resp) {
                 if (gpg::IsSuccess(resp.status)) {
                     CCLOG("rematch success");
                 } else {
                     CCLOG("rematch failed:%d", resp.status);
                 }
             });
}

void TurnBasedMultiplayerScene::LeaveMatch(cocos2d::Ref *sender) {
    if (!_match.Valid()) {
        CCLOG("Match is invalid");
        return;
    }
    
    auto& tbmp =_game_services->TurnBasedMultiplayer();
    if (gpg::MatchStatus::MY_TURN == _match.Status()) {
        tbmp.LeaveMatchDuringMyTurn(_match,
                                _match.SuggestedNextParticipant(),
                                [](gpg::MultiplayerStatus status) {
                                    if (gpg::IsSuccess(status)) {
                                        CCLOG("leave match during my turn success");
                                    } else {
                                        CCLOG("leave match during my turn failed");
                                    }
                                });
    } else if (gpg::MatchStatus::THEIR_TURN == _match.Status()) {
        tbmp.LeaveMatchDuringTheirTurn(_match,
                               [](gpg::MultiplayerStatus status) {
                                   if (gpg::IsSuccess(status)) {
                                       CCLOG("leave match during their turn success");
                                   } else {
                                       CCLOG("leave match during their turn failed");
                                   }
                               });
    } else {
        CCLOG("leave match staus unandle:%d", _match.Status());
    }
}

void TurnBasedMultiplayerScene::CancelMatch(cocos2d::Ref *sender) {
    if (!_match.Valid()) {
        CCLOG("Match is invalid");
        return;
    }
    
    _game_services->TurnBasedMultiplayer()
    .CancelMatch(_match,
                 [](gpg::MultiplayerStatus status) {
                     if (gpg::IsSuccess(status)) {
                         CCLOG("cancel match success");
                     } else {
                         CCLOG("cancel match failed");
                     }
                 });
}

void TurnBasedMultiplayerScene::FetchMatches(cocos2d::Ref *sender) {
    if (!_match.Valid()) {
        CCLOG("Match is invalid");
        return;
    }
    
    _game_services->TurnBasedMultiplayer()
    .FetchMatches([](gpg::TurnBasedMultiplayerManager::TurnBasedMatchesResponse const & resp) {
        if (gpg::IsSuccess(resp.status)) {
            CCLOG("fetch matches success");
        } else {
            CCLOG("fetch matches failed:%d", resp.status);
        }
    });
}

void TurnBasedMultiplayerScene::handleMatch(gpg::TurnBasedMatch match) {
    switch (match.Status()) {
        case gpg::MatchStatus::INVITED: {
            break;
        }
        case gpg::MatchStatus::THEIR_TURN: {
            break;
        }
        case gpg::MatchStatus::MY_TURN: {
            this->_myId = _match.PendingParticipant().Id();
            break;
        }
        case gpg::MatchStatus::PENDING_COMPLETION: {
            break;
        }
        case gpg::MatchStatus::COMPLETED: {
            break;
        }
        case gpg::MatchStatus::CANCELED: {
            break;
        }
        case gpg::MatchStatus::EXPIRED: {
            break;
        }
        default: {
            break;
        }
    }
}

File: js/src/TurnBasedMultiplayerScene.js
var LayerTurnBasedMultiplayer = cc.Layer.extend({

    _lbInfo: null,

    ctor: function() {
        //////////////////////////////
        // 1. super init first
        this._super();

        this.__createMenu();

        this._lbInfo = new cc.LabelTTF("", "sans", 24);
        this._lbInfo.x = cc.Director.getInstance().getWinSize().width / 2;
        this._lbInfo.y = 120;
        this.addChild(this._lbInfo);
    },

    __createMenu: function() {

        var self = this;

        cc.MenuItemFont.setFontName("sans");
        var size = cc.Director.getInstance().getWinSize();

        var menu = new cc.Menu(

            new cc.MenuItemFont("Create TurnBased Match", function() {
                let params = {
                    type: "ui", // "quick_match"
                    minimumAutomatchingPlayers: 1,
                    maximumAutomatchingPlayers: 2
                    // playerIds: 0
                };
                _context.gameServices().TurnBasedMultiplayer.CreateTurnBasedMatch(function(result) {
                    if (gpg.IsSuccess(result.result)) {
                        cc.log('Create TurnBased Match Success');
                        self._match = result.match;
                        self._myId = result.match.pendingParticipant.id;
                        self.handleMatch(self._match);
                    } else {
                        cc.log('Create TurnBased Match Failed:' + result.result);
                    }
                }, params);
            }),
            new cc.MenuItemFont("Show Match Inbox UI", function() {
                _context.gameServices().TurnBasedMultiplayer.ShowMatchInboxUI(function(result) {
                    if (gpg.IsSuccess(result.result)) {
                        cc.log('Show Match Inbox UI Success');
                        self._match = result.match;
                        self.handleMatch(self._match);
                    } else {
                        cc.log('Show Match Inbox UI Failed');
                    }
                });
            }),
            new cc.MenuItemFont('Dismiss Match', function() {
                if (null == self._match) {
                    cc.log('Match is none');
                    return
                }
                _context.gameServices().TurnBasedMultiplayer.DismissMatch(self._match.id);
            }),
            new cc.MenuItemFont('Fetch Match', function() {
                if (null == self._match) {
                    cc.log('Match is none');
                    return
                }
                _context.gameServices().TurnBasedMultiplayer.FetchMatch(function(result) {
                    if (gpg.IsSuccess(result.result)) {
                        cc.log('Fetch Match Success');
                    } else {
                        cc.log('Fetch Match Failed');
                    }
                }, self._match.id);
            }),
            new cc.MenuItemFont('Take My Turn', function() {
                if (null == self._match) {
                    cc.log('Match is none');
                    return
                }
                let participant_results_id = self._match.pendingParticipant.id;
                _context.gameServices().TurnBasedMultiplayer.createParticipantResult(self._myId,
                    participant_results_id,
                    0,
                    gpg.MatchResult.LOSS);
                let next_participant_id = self._match.suggestedNextParticipant.id;
                let data = JSON.stringify({counter: 1});
                _context.gameServices().TurnBasedMultiplayer.TakeMyTurn(function(result) {
                    if (gpg.IsSuccess(result.result)) {
                        cc.log('Take My Turn Success');
                    } else {
                        cc.log('Take My Turn Failed');
                    }
                },
                self._match.id,
                participant_results_id,
                next_participant_id,
                data);
            }),
            new cc.MenuItemFont('Finish Match During MyTurn', function() {
                if (null == self._match) {
                    cc.log('Match is none');
                    return
                }
                let participant_results_id = self._match.pendingParticipant.id;
                let participant_results_id = self._match.pendingParticipant.id;
                _context.gameServices().TurnBasedMultiplayer.createParticipantResult(self._myId,
                    participant_results_id,
                    0,
                    gpg.MatchResult.WIN);
                let data = JSON.stringify({counter: 1})
                _context.gameServices().TurnBasedMultiplayer.FinishMatchDuringMyTurn(function(result) {
                    if (gpg.IsSuccess(result.result)) {
                        cc.log('Finish Match During MyTurn Success');
                    } else {
                        cc.log('Finish Match During MyTurn Failed');
                    }
                },
                self._match.id,
                participant_results_id,
                data);
            }),
            new cc.MenuItemFont('Rematch', function() {
                if (null == self._match) {
                    cc.log('Match is none');
                    return
                }
                _context.gameServices().TurnBasedMultiplayer.Rematch(function(result) {
                    if (gpg.IsSuccess(result.result)) {
                        cc.log('Rematch Success');
                    } else {
                        cc.log('Rematch Failed');
                    }
                },
                self._match.id);
            }),
            new cc.MenuItemFont('Leave Match', function() {
                if (null == self._match) {
                    cc.log('Match is none');
                    return
                }
                if (self._match.matchStatus == gpg.MatchStatus.MY_TURN) {
                    _context.gameServices().TurnBasedMultiplayer.LeaveMatchDuringMyTurn(function(result) {
                        if (gpg.IsSuccess(result.result)) {
                            cc.log('Leave Match Success');
                        } else {
                            cc.log('Leave Match Failed');
                        }
                    },
                    self._match.id,
                    self._match.suggestedNextParticipant.id);
                } else {
                    _context.gameServices().TurnBasedMultiplayer.LeaveMatchDuringTheirTurn(function(result) {
                        if (gpg.IsSuccess(result.result)) {
                            cc.log('Leave Match Success');
                        } else {
                            cc.log('Leave Match Failed');
                        }
                    },
                    self._match.id);
                }
            }),
            new cc.MenuItemFont('Cancel Match', function() {
                if (null == self._match) {
                    cc.log('Match is none');
                    return
                }
                _context.gameServices().TurnBasedMultiplayer.CancelMatch(function(result) {
                    if (gpg.IsSuccess(result.result)) {
                        cc.log('Cancel Match Success');
                    } else {
                        cc.log('Cancel Match Failed');
                    }
                },
                self._match.id);
            }),
            new cc.MenuItemFont('Fetch Matches', function() {
                _context.gameServices().TurnBasedMultiplayer.FetchMatches(function(result) {
                    if (gpg.IsSuccess(result.result)) {
                        cc.log('Fetch Match Success');
                    } else {
                        cc.log('Fetch Match Failed');
                    }
                });
            }),
            new cc.MenuItemFont('Show Player Select UI', function() {
                let min_match_player = 1;
                let max_match_player = 2;
                let allow_automatch = true;
                _context.gameServices().TurnBasedMultiplayer.ShowPlayerSelectUI(function(result) {
                    if (gpg.IsSuccess(result.result)) {
                        cc.log('Show Player Select UI Success');
                    } else {
                        cc.log('Show Player Select UI Failed');
                    }
                },
                min_match_player,
                max_match_player,
                allow_automatch);
            }),

            new cc.MenuItemFont("Main menu", function() {
                cc.director.runScene(new HelloWorldScene());
            })
        );

        menu.alignItemsVerticallyWithPadding(3);
        menu.x = size.width / 2;
        menu.y = size.height / 2 + 40;
        this.addChild(menu);
    },

    handleMatch: function(match) {
        if (match.matchStatus == gpg.MatchStatus.MY_TURN) {
            this._myId = match.pendingParticipant.id;
        } else if (match.matchStatus == gpg.MatchStatus.THEIR_TURN) {
        } else if (match.matchStatus == gpg.MatchStatus.COMPLETED) {
        } else if (match.matchStatus == gpg.MatchStatus.EXPIRED) {
        } else {
        }
    }

});

var S_TurnBasedMultiplayer = cc.Scene.extend({
    onEnter: function() {
        this._super();
        this.addChild(new LayerTurnBasedMultiplayer());
    }
});
